includes/class-wc-coupon.php

```diff
@@ -54,97 +78,241 @@ class WC_Coupon {
 	const WC_COUPON_SUCCESS                          = 200;
 	const WC_COUPON_REMOVED                          = 201;
 
-	/** @public string Coupon code. */
-	public $code   = '';
+	/**
+	 * Coupon constructor. Loads coupon data.
+	 * @param  mixed $code code of the coupon to load
+	 */
+	public function __construct( $code = '' ) {
+		if ( $code instanceof WC_Coupon ) {
+			$this->read( absint( $code->get_id() ) );
+		} elseif ( $coupon = apply_filters( 'woocommerce_get_shop_coupon_data', false, $code ) ) {
```

Comment:
Let's deprecate it first. I know at least Points & Rewards is using it (which we have control over) but I'm sure there are other ones, and it would probably be good to give a version to switch.